rskj-core/src/main/java/org/ethereum/util/RLP.java

```diff
@@ -191,43 +199,62 @@ public static BigInteger decodeBigInteger(byte[] data, int index) {
         return BigIntegers.fromUnsignedByteArray(bytes);
     }
 
+    private static void assertLengthOfLength(byte[] payload,byte lengthOfLength) {
+        if (lengthOfLength >= payload.length) {
+            throw new RLPException("lengthOfLength cannot be higher than payload length");
+        }
+    }
+
     public static int getNextElementIndex(byte[] payload, int pos) {
 
         if (pos >= payload.length) {
             return -1;
         }
 
+        int nextElementIndex = -1;
+
         if ((payload[pos] & 0xFF) >= OFFSET_LONG_LIST) {
             byte lengthOfLength = (byte) (payload[pos] - OFFSET_LONG_LIST);
+
+            assertLengthOfLength(payload, lengthOfLength);
+
             int length = calcLength(lengthOfLength, payload, pos);
-            return pos + lengthOfLength + length + 1;
+            nextElementIndex = pos + lengthOfLength + length + 1;
         }
         if ((payload[pos] & 0xFF) >= OFFSET_SHORT_LIST
                 && (payload[pos] & 0xFF) < OFFSET_LONG_LIST) {
 
             byte length = (byte) ((payload[pos] & 0xFF) - OFFSET_SHORT_LIST);
-            return pos + 1 + length;
+            nextElementIndex = pos + 1 + length;
         }
         if ((payload[pos] & 0xFF) >= OFFSET_LONG_ITEM
                 && (payload[pos] & 0xFF) < OFFSET_SHORT_LIST) {
 
             byte lengthOfLength = (byte) (payload[pos] - OFFSET_LONG_ITEM);
+
+            assertLengthOfLength(payload, lengthOfLength);
+
             int length = calcLength(lengthOfLength, payload, pos);
-            return pos + lengthOfLength + length + 1;
+            nextElementIndex = pos + lengthOfLength + length + 1;
         }
         if ((payload[pos] & 0xFF) > OFFSET_SHORT_ITEM
                 && (payload[pos] & 0xFF) < OFFSET_LONG_ITEM) {
 
             byte length = (byte) ((payload[pos] & 0xFF) - OFFSET_SHORT_ITEM);
-            return pos + 1 + length;
+            nextElementIndex = pos + 1 + length;
         }
         if ((payload[pos] & 0xFF) == OFFSET_SHORT_ITEM) {
-            return pos + 1;
+            nextElementIndex = pos + 1;
         }
         if ((payload[pos] & 0xFF) < OFFSET_SHORT_ITEM) {
-            return pos + 1;
+            nextElementIndex = pos + 1;
```

Comment:
With these changes we are not breaking the if check, so we ware always checking of the if conditionals. I think it would be better to avoid it. 
We can avoid id by using else if or extracting this part of get the nextElementIndex to a functio nand continue with the other checks later.

Reply:
That makes sense! Updating the PR